### PR TITLE
Merging CUDA API into dev 

### PR DESCRIPTION
**Pull request by bornaehsani**
This pull request merges the CUDA API into f1 dev. 

Summary of modifications: 

1. Added bsg_manycore_cuda.cpp/h into the cl_manycore f1 libraries. This file includes all host primitives and routines responsible for
     - Initializing a device with a pool of tiles
     - Loading the binary on all tiles
     - Allocating grids of tile groups on the network
     - Allocating memory on device DRAM, and moving memory to and from device DRAM
     - Sending packets to tiles setting their __bsg_x/y, __bsg_id, __bsg_tile_group_id_x/y, __bsg_grp_org_x/y, CSR_TGO_X/Y, __bsg_grid_dim_x/y symbols. 
     - Sending packets to tiles setting the argc, argv, finish signal address and kernel pointer. 
     - Scheduling grids of tile groups to launch and execute on the device
     - Synchronizing with tile groups to wait for their finish packet

2. Separated the spmd and cuda regression test directories. Added the following regression tests for cuda. Each test has a corresponding directory in the dev_cuda_v4 (to be master) branch of the bsg_manycore/software/spmd/bsg_cuda_lite_runtime. 
     - test_empty: Launches a 1x1 grid of 2x2 tile group to run the empty kernel.
     - test_empty_parallel: Launches a 4x2 grid of 2x2 tile groups to all run the empty kernel.
     - test_barrier: Launches a 1x1 grid of 2x2 tile group to run the barrier kernel.
     - test_packet: Launches a 1x1 grid of 2x2 tile group to run the packet kernel. Each tile sends packets with all the symbols that the CUDA API is charged with setting, and the host API picks them up to  validate their value.
     - test_stack_load: Launches a 1x1 grid of 2x2 tile group to load 16 arguments (8 go through stack) and store the sum into DRAM. The host then picks up the sum to compare and validate stack usability.
     - test_shared_mem: Launches a 1x1 grid of 2x2 tile group to run the shared_mem kernel. Tiles in a  tile group declare a tile group shared memory, store to it and load with different striping algorithms to evaluate the function of barrier and shared memory.
     - test_shared_mem_load_store: Launches a grid of multiple tile groups to run the shared_mem_load_store kernel. Host loads a MxN matrix into device DRAM, each tile loads a portion into tile group shared memory and stores to another location. Host the picks up the data and compares.
     - test_vec_add: Launches a 1x1 grid of 2x2 tile group to run vec_add kernel. Host loads two arrays into device dram and picks up the result to compare.
     - test_vec_add_parallel: Launches a grid of multiple 2x2 tile groups to perform the vec_add_parallel kernel. Each tile group hands a portion of the vector addition.
     - test_vec_add_parallel_multi_grid: Launches two different grids of 2x2 tile groups to perform vector addition on two different vectors separately.
     - test_matrix_mul: Launches a grid of multiple 2x2 tile groups to perform matrix_mul kernel. Each tile group handles a portion  (block_size_y * block_size_x) of the result matrix.
     - test_matrix_mul_shared_mem: Launches a grid of multiple 2x2 tile groups to perform matrix_mul with shared memory. Not working.
All cuda regression tests are heavily documented and include commentary for each section of the code.

Regression tests passed for all library/spmd/cuda regression tests. Rebased, no conflict, and ready to merge.

### CODE REVIEW

**Comment by max-ruttenberg** [comment]
mother of god.

File: cl_manycore/libraries/bsg_manycore_driver.cpp [hunk 1]
```diff
@@ -434,6 +434,10 @@ int hb_mc_fifo_transmit (uint8_t fd, hb_mc_fifo_tx_t type, const hb_mc_packet_t
 	// Write 1 to the Transmit Complete bit to clear it
 	hb_mc_fifo_set_ixr_bit(fd, dir, HB_MC_MMIO_FIFO_ISR_OFFSET, HB_MC_MMIO_FIFO_IXR_TC_BIT);
 
+	#ifdef DEBUG
```
Comment (max-ruttenberg) [inline cl_manycore/libraries/bsg_manycore_driver.cpp]: there are print macros for doing this thing in `bsg_manycore_printing.h`. please use them.

**Comment by bornaehsani** [comment]
My apologies for the size of the pull request. Upkeep and maintainability was handled poorly on my part. I tried to compensate with commentary in the files.

**Comment by max-ruttenberg** [comment]
NP, I'm just teasing.

File: cl_manycore/libraries/bsg_manycore_tile.cpp [hunk 2]
```diff
@@ -55,51 +60,259 @@ int hb_mc_tile_unfreeze (uint8_t fd, uint8_t x, uint8_t y) {
                                     x, y, HB_MC_PACKET_OP_REMOTE_STORE);
 	if (hb_mc_fifo_transmit(fd, HB_MC_FIFO_TX_REQ, &unfreeze) != HB_MC_SUCCESS)
 		return HB_MC_FAIL;
-	else
-		return HB_MC_SUCCESS;
+
+	#ifdef DEBUG
```
Comment (max-ruttenberg) [inline cl_manycore/libraries/bsg_manycore_tile.cpp]: please use the print macros defined in `bsg_manycore_printing.h`

**Comment by bornaehsani** [comment]
Oh no dude you're absolutely right its huge! I have to learn to merge more frequently =)

File: cl_manycore/libraries/bsg_manycore_tile.cpp (hunk 2 above)
Comment (bornaehsani) [reply cl_manycore/libraries/bsg_manycore_tile.cpp]: This is great! I will change all these statements, thanks for pointing it out!

File: cl_manycore/libraries/bsg_manycore_tile.cpp [hunk 3]
```diff
@@ -55,51 +60,259 @@ int hb_mc_tile_unfreeze (uint8_t fd, uint8_t x, uint8_t y) {
...
+		return HB_MC_FAIL;
+	}
+
+	if (symbol_to_eva(elf, "__bsg_grid_dim_y", &bsg_grid_dim_y_eva) != HB_MC_SUCCESS){
+		fprintf(stderr, "hb_mc_tile_set_grid_dim_symbols() --> symbol_to_eva(): failed to aquire __bsg_grid_dim_y eva.\n");
+		return HB_MC_FAIL;
+	}
+
+	if (hb_mc_copy_to_epa(fd, x, y, bsg_grid_dim_x_eva >> 2 /* TODO: magic number */, &grid_dim_x, 1) != HB_MC_SUCCESS) {
+		fprintf(stderr, "hb_mc_tile_set_grid_dim_symbols() --> hb_mc_copy_to_epa(): failed to set tile __bsg_grid_dim_x symbol.\n");
```
Comment (max-ruttenberg) [inline cl_manycore/libraries/bsg_manycore_tile.cpp]: * I recommend using the automatic variable `__func__`  instead of hard coding the string.

Comment (bornaehsani) [reply cl_manycore/libraries/bsg_manycore_tile.cpp]: Good point! will do, thanks!

**Pull request closed without merging**